Implemented some series methods for gamma and error functions

Abhishekjsr283204 opened the pull request:
References to other Issues or PRs

Brief description of what is fixed or changed

While implementing the leading term for expint(nu, z) function, we don't have to take care of cases where nu < 1, and as the already implemented _eval_nseries method shows we essentially only need to focus on cases where nu >=1.
Because expint(nu, z) when nu < 1 will be rewritten in the forms of exponentials.


Other comments

<!-- BEGIN RELEASE NOTES -->
* series
  * Implemented leading term methods for expint and Si function.
  * Implemented leading term methods for uppergamma and lowergamma function.
<!-- END RELEASE NOTES -->

Abhishekjsr283204 committed:
remove whitespaces

Abhishekjsr283204 committed:
Implemented some series methods for gamma and error functions

VascoSch92 commented on sympy/functions/special/error_functions.py:
```diff
@@ -1409,7 +1409,19 @@ def _eval_rewrite_as_Si(self, nu, z, **kwargs):
+        nu, z = self.args
+        if nu.is_real and nu >= 1:
+            term_one = (-z)**(nu - 1)/factorial(nu - 1) * (digamma(nu) - log(z))
+            term_two = 1/(nu - 1)
```
you can have division by zero here, right?
You need something like that:
```python
if nu == 1:
     term_two = z/(2 - nu)
else:
    term_two = 1/(nu - 1)
```

Abhishekjsr283204 replied:
Thank you for your review. You raised a valid concern about the potential for division by zero in the calculation of term_two. I agree that this could be a problem in certain cases. To address this issue, I'll ensure that nu - 1 is not equal to zero before performing the division. If nu equals 1, I'll handle the calculation differently to avoid division by zero

DarkNisa replied:
some thoughts over the code:
`nu == 1:` the problem here that `term_two = z/(2 - nu)`
is confusing statement and better to be rewritten into `term_two = z`
and also it is confusing having intersected nested `if` statements
`nu >= 1` and `nu == 1` so I think it's better to rewrite it more clearly without such intersections
`arg` variable is unnecessary here and better to be moved to `return` statement
like `return Add(*[term_one, term_two]).as_leading_term(x, logx=logx)`

anutosh491 replied:
Correct please add tests for the same (and also add the corner cases being questioned here). It shall help us review much more earily

Abhishekjsr283204 replied:
Yes! you are correct i will fix it

Abhishekjsr283204 committed:
Pass the code quality

Abhishekjsr283204 committed:
remove whitespace in gamma

DarkNisa commented:
So the main problem here is that you haven't written pull request properly
so it doesn't start testing the whole changes, do you need to rewrite it according 
the guide that sympy bot attached earlier.

DarkNisa commented on sympy/functions/special/error_functions.py:
```diff
@@ -1895,6 +1907,19 @@ def _eval_rewrite_as_sinc(self, z, **kwargs):
+        arg0 = arg.subs(x, 0)
+
+        if arg0 is S.NaN:
+            arg0 = arg.limit(x, 0, dir='-' if re(cdir).is_negative else '+')
```
some thoughts over the code:
`arg0 = arg.limit(x, 0, dir='-' if re(cdir).is_negative else '+')`
maybe it's better to move dir calculation to temp variable to make it more readable

Abhishekjsr283204 committed:
Merge branch 'master' into JSR

Abhishekjsr283204 committed:
Eliminate Whitespace in limit file

Abhishekjsr283204 committed:
Merge branch 'JSR' of https://github.com/Abhishekjsr283204/sympy into JSR

anutosh491 commented on sympy/functions/special/gamma_functions.py:
```diff
@@ -383,12 +383,18 @@ def _eval_aseries(self, n, args0, x, logx):
 
+    def _eval_as_leading_term(self, x, logx=None, cdir=0):
+        return self.rewrite(uppergamma)._eval_as_leading_term(x, logx=logx, cdir=cdir)
+
```
Add tests for this.

anutosh491 commented on sympy/functions/special/gamma_functions.py:
```diff
@@ -556,6 +562,10 @@ def _eval_rewrite_as_expint(self, s, x, **kwargs):
+    def _eval_as_leading_term(self, x, logx=None, cdir=0):
+        from sympy.functions.special.error_functions import expint
+        return self.rewrite(expint)._eval_as_leading_term(x, logx=logx, cdir=cdir)
+
```
Add tests for this too.

anutosh491 commented:
Also please fix the code quality. I think you might have to pull the latest changes and rebase on top of it.

Abhishekjsr283204 committed:
fix code quality

Abhishekjsr283204 committed:
Add the credential

Abhishekjsr283204 committed:
ADD MAILMAP

Abhishekjsr283204 committed:
mailmap

author: add Abhishek kumar to .mailmap

ADDED NAME

git commit -m "squash commit"

Abhishekjsr283204 committed:
fix the test

Abhishekjsr283204 committed:
fix the reordered

Abhishekjsr283204 committed:
author: add Abhishek Kumar to .mailmap

Abhishekjsr283204 committed:
Implemented the solve method for the cable class

anutosh491 commented on .mailmap:
```diff
@@ -49,9 +49,9 @@
+#   https://docs.sympy.org/dev/contributing/new-contributors-guide/workflow-.html#add-your-name-and-email-address-to-the-mailmap-file
 #
-#   ...
+#   ...process
```
I don't understand these changes in the mailmap file. I am not sure we need them

Abhishekjsr283204 replied:
Oops i will fix it

anutosh491 commented on .mailmap:
```diff
@@ -185,6 +185,8 @@
 # See also: 'git shortlog --help' and 'git check-mailmap --help'.
 #
+
+
```
Why do we need these lines here ?

Abhishekjsr283204 replied:
Yes, this is the unnecessary line i will fix it!

Abhishekjsr283204 committed:
correction in mailmap

Abhishekjsr283204 committed:
remove the unnecessary the =
